Updates series.py with midding sum_series module

ericstreit opened the pull request:
I couldn't come up with a working recursive solution for sum_series so I did it iteratively.

PythonCHB commented:
this will get you credit -- but do try to look at the recursive approach.

recursion means calling a function from within itself.

read this again:

https://uwpce-pythoncert.github.io/PythonCertDevel/modules/Recursion.html

(I've added more to make it more clear)

And maybe start with fibonacci series, and once you have that, then try the generic sum_series

PythonCHB commented:
OK -- even though much of this is already in the repo -- it' showing a bunch of new files -- any idea what's up???

I"m not going to merge 'till we have that figured out.

ericstreit commented:
Ah. Originally I had created a "lesson02" folder and commited/pushed and
pull requestd on it. Afterwards I changed the folder to "session02" as per
the instructions and then I think I copied over the files from lesson02 to
session02. I removed lesson02 from git and pushed. Everything looks good on
my repo.

On Fri, Oct 19, 2018 at 11:02 PM Christopher H.Barker, PhD <
notifications@github.com> wrote:

> OK -- even though much of this is already in the repo -- it' showing a
> bunch of new files -- any idea what's up???
>
> I"m not going to merge 'till we have that figured out.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/UWPCE-PythonCert-ClassRepos/Fall2018-PY210A/pull/58#issuecomment-431552670>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AqE_F4ixn7kCUCvmpwkWmQx6cXg7C4Fwks5umryGgaJpZM4Xxa4f>
> .
>

PythonCHB commented on students/ericstreit/.gitignore:
```diff
@@ -0,0 +1,106 @@
+.DS_Store
```
you don't most (any?) of these, 'cuase they are in the main .gitignore.

but it won't hurt, either.

PythonCHB commented on students/ericstreit/session02/series.py:
```diff
@@ -0,0 +1,82 @@
+    """This is the recursive sum series function"""
+    if n == n2: return n2
+    elif n == 1: return 1
+    else: return sum_seriesr(n-1)+sum_seriesr(n-2)
```
you need to pass n2 and n3 on through

PythonCHB commented on students/ericstreit/session03/list_lab.py:
```diff
@@ -0,0 +1,79 @@
+print("The list currently now contains: ", mylist)
+l = len(mylist)
+listitem = int(input("Enter a list item from 1 to {} and I will show it to you: ".format(l)))
+print(mylist[(listitem-1)])
```
no need for extra parentheses here.

PythonCHB commented on students/ericstreit/session03/list_lab.py:
```diff
@@ -0,0 +1,79 @@
+print()
+pause = input("OK, let's add Kiwis to the list!")
+newlist = ["Kiwis"]
+mylist = newlist + mylist
```
you could use += here -- but better yet, use `mylist.append("KIwis")`

PythonCHB commented on students/ericstreit/session03/list_lab.py:
```diff
@@ -0,0 +1,79 @@
+mylist.insert(0,"Mangos")
+print(mylist)
+pause = input("Now displaying all items that begin with the letter 'P'")
+for i in mylist:
```
I tend to use more meaning full names that "i" -- i is good for indexes, but when the loop is more than two lines, it helps have a meaning ful name:

```
for fruit in mylist:
   ...
```

PythonCHB commented on students/ericstreit/session03/list_lab.py:
```diff
@@ -0,0 +1,79 @@
+    elif choice == "n":
+         mylist.remove(i)
+         print("Removing {} from the list!".format(i))
+         #bug if 'n' is selected it skips the next item in the list
```
yup -- this is a common "gotcha" with mutable sequences (i.e. lists...):

if you mutate the list, particular if you remove something, the looping gets messed up -- "where am I in the list?"

so you want to loop through a copy, and mutate the original:

```
for i in mylist[:]:
   ...
```

PythonCHB commented on students/ericstreit/session03/list_lab.py:
```diff
@@ -0,0 +1,79 @@
+
+#for testing
+if __name__=="__main__":
+    pass
```
why an if __name__ block with nothing in it?

PythonCHB commented on students/ericstreit/session03/slicing.py:
```diff
@@ -0,0 +1,44 @@
+def numberswap(n):
+    """Prints the list with the last item first and first item last"""
+    l=len(n)
+    print("{}, {} and {}".format(n[-1], n[1:(l-1)], n[0]))
```
correct, but the idea was to make a new sequence, not just print it :-)

PythonCHB commented on students/ericstreit/session03/slicing.py:
```diff
@@ -0,0 +1,44 @@
+
+def every_other_in_middle(n):
+    """Prints every other item in list omitting the first and last 4 in the series"""
+    l=len(n)
```
you hardly ever need to use the length -- try using a negative index

PythonCHB commented on students/ericstreit/session03/slicing.py:
```diff
@@ -0,0 +1,44 @@
+
+def reverse_list(n):
+    """Prints the reverse of the list"""
+    print(n[::-1])
```
yup!

PythonCHB commented on students/ericstreit/session03/slicing.py:
```diff
@@ -0,0 +1,44 @@
+
+#Function testing
+if __name__=="__main__":
+    mylist = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15]
```
if your functions returned the new sequence, then these could have been asserts checking the correct answer!

PythonCHB commented on students/ericstreit/session03/str_format_lab.py:
```diff
@@ -0,0 +1,61 @@
+    """string exercise task 3"""
+    l = len(n)
+    print(("the {} numbers are: " + ",".join(["{}"] * l).format(l, *n)))
+
```
sometimes it's easier to put things on separate lines :-)

```
mytup = (1, 2, 3, 4, 5)


def formatter(n):
    """string exercise task 3"""
    l = len(n)
    format_str = "the {} numbers are: " + ",".join(["{}"] * l)
    return format_str.format(l, *n)

print(formatter(mytup))
```
You had an end parens in the wrong place -- so it was applying teh format to the second part of your string, not the whole thing...

PythonCHB commented:
You have conflicts in series.py -- take a look at the PR in gitHUb and see if you can figure out how to resolve them.

Not sure how it happened -- you are teh only one chaning your code -- odd.

You could also try doing:

git pull upstream master
git push

and that _might_ synchronize it.

PythonCHB commented on students/ericstreit/session03/mailroom.py:
```diff
@@ -3,16 +3,14 @@
 donor3 = ("Hester Shaw", [5, 92, 101])
 donor4 = ("Katherine Valentine", [1000, 2000, 3000])
 donor5 = ("Grike", [1])
 donors = [donor1] + [donor2] + [donor3] + [donor4] + [donor5]
```
you can build a list by just putting the items in the list:
```
donors = [donor1, donor2, donor3, donor4, donor5]
```

PythonCHB commented on students/ericstreit/session03/mailroom.py:
```diff
@@ -32,45 +32,54 @@ def mainmenu():
 
 def compose_email(name, amount):
     """This function will compose the thank you email"""
     print("\n\nEMAIL SENT:\n'Thank you, {}, for your generous contribution of ${}! We will be sure to ask you for more money again soon.'\n\n".format(name, amount))
```
a good idea to break this down into more than one line -- not great to have a line this long...

you could even define the letter template outside his function altogether -- maybe with a triple quoted string.

PythonCHB commented on students/ericstreit/session03/mailroom.py:
```diff
@@ -32,45 +32,54 @@ def mainmenu():
+#    l = len(donors)
+#    print(name)
+#variable if a name is found in the search. default is 0 / False
     found = 0
```
while 0 is falsey -- better to use the actual False if that's what you mean.

PythonCHB commented on students/ericstreit/session03/mailroom.py:
```diff
@@ -32,45 +32,54 @@ def mainmenu():
     found = 0
+#for loop to search for a donor name in the database. If found it will flag the found variable as True and add the addtional amount
     for i in range(len(donors)):
         if name in donors[i]:
```
better to search the donor list for the donor on its own, and then pass the result out (maybe even a separat function) but not put the manipulation logic right in that loop.

PythonCHB commented on students/ericstreit/session03/mailroom.py:
```diff
@@ -32,45 +32,54 @@ def mainmenu():
         if name in donors[i]:
-            amount = int(input("How much $$ did they donate this time?: "))
+            print("How much $$ did they donate this time?: ")
+            amount = int(input(": "))
```
do they have to donate in integer number of dollars? no $2.50?  -- your choice, but do make it clear. Any reason not to use float(input...)?

PythonCHB commented on students/ericstreit/session03/mailroom.py:
```diff
@@ -32,45 +32,54 @@ def mainmenu():
             break
         else:
             continue
     if not found:
```
you can use a "else" block with the for loop instead of this "sentinel" value?

https://uwpce-pythoncert.github.io/PythonCertDevel/modules/Iteration.html#else

ericstreit commented:
Thanks for the feedback!  I'll work on that this weekend.

I also didn't realize that you were pulling these yet, when I "commit" and
push to my repo, it is not notifying you is it? I was under the impression
that you only pulled to the master after I made a pull request or perhaps I
understood that workflow incorrectly? I ask because I commit often to save
my work but don't want to pepper you with requests when the changes are
very minor.

Eric

On Fri, Oct 26, 2018 at 7:15 PM Christopher H.Barker, PhD <
notifications@github.com> wrote:

> *@PythonCHB* commented on this pull request.
>
> looking really good -- almost there!
> ------------------------------
>
> In students/ericstreit/session03/mailroom.py
> <https://github.com/UWPCE-PythonCert-ClassRepos/Fall2018-PY210A/pull/58#discussion_r228698063>
> :
>
> > @@ -3,16 +3,14 @@
>  #
>  #!/usr/bin/env python3
>
> -#define variables - the donor list
> +#define variables - create donor list
>  donor1 = ("Anna Fang", [23, 5000])
>  donor2 = ("Tom Natsworthy", [99, 783, 3])
>  donor3 = ("Hester Shaw", [5, 92, 101])
>  donor4 = ("Katherine Valentine", [1000, 2000, 3000])
>  donor5 = ("Grike", [1])
>  donors = [donor1] + [donor2] + [donor3] + [donor4] + [donor5]
>
> you can build a list by just putting the items in the list:
>
> donors = [donor1, donor2, donor3, donor4, donor5]
>
> ------------------------------
>
> In students/ericstreit/session03/mailroom.py
> <https://github.com/UWPCE-PythonCert-ClassRepos/Fall2018-PY210A/pull/58#discussion_r228698143>
> :
>
> >      else:
> -        donorupdate(name)
> +        donor_update(name)
> +
> +def list_users():
> +    """A simple function that lists donor names"""
> +    print("\nDonors\n", "{:-<60}".format(""))
> +    for i in donors:
> +        print(i[0])
>
>  def compose_email(name, amount):
>      """This function will compose the thank you email"""
>      print("\n\nEMAIL SENT:\n'Thank you, {}, for your generous contribution of ${}! We will be sure to ask you for more money again soon.'\n\n".format(name, amount))
>
> a good idea to break this down into more than one line -- not great to
> have a line this long...
>
> you could even define the letter template outside his function altogether
> -- maybe with a triple quoted string.
> ------------------------------
>
> In students/ericstreit/session03/mailroom.py
> <https://github.com/UWPCE-PythonCert-ClassRepos/Fall2018-PY210A/pull/58#discussion_r228698165>
> :
>
> >      """This function will update the donor list with a new donor or a new amount given by a current donor"""
> -    l = len(donors)
> -    print(name)
> +#    l = len(donors)
> +#    print(name)
> +#variable if a name is found in the search. default is 0 / False
>      found = 0
>
> while 0 is falsey -- better to use the actual False if that's what you
> mean.
> ------------------------------
>
> In students/ericstreit/session03/mailroom.py
> <https://github.com/UWPCE-PythonCert-ClassRepos/Fall2018-PY210A/pull/58#discussion_r228698189>
> :
>
> >      found = 0
> +#for loop to search for a donor name in the database. If found it will flag the found variable as True and add the addtional amount
>      for i in range(len(donors)):
>          if name in donors[i]:
>
> better to search the donor list for the donor on its own, and then pass
> the result out (maybe even a separat function) but not put the manipulation
> logic right in that loop.
> ------------------------------
>
> In students/ericstreit/session03/mailroom.py
> <https://github.com/UWPCE-PythonCert-ClassRepos/Fall2018-PY210A/pull/58#discussion_r228698218>
> :
>
> >      for i in range(len(donors)):
>          if name in donors[i]:
> -            amount = int(input("How much $$ did they donate this time?: "))
> +            print("How much $$ did they donate this time?: ")
> +            amount = int(input(": "))
>
> do they have to donate in integer number of dollars? no $2.50? -- your
> choice, but do make it clear. Any reason not to use float(input...)?
> ------------------------------
>
> In students/ericstreit/session03/mailroom.py
> <https://github.com/UWPCE-PythonCert-ClassRepos/Fall2018-PY210A/pull/58#discussion_r228698254>
> :
>
> >      for i in range(len(donors)):
>          if name in donors[i]:
> -            amount = int(input("How much $$ did they donate this time?: "))
> +            print("How much $$ did they donate this time?: ")
> +            amount = int(input(": "))
>              donors[i][1].append(amount)
>              found = 1
>              break
>          else:
>              continue
>      if not found:
>
> you can use a "else" block with the for loop instead of this "sentinel"
> value?
>
>
> https://uwpce-pythoncert.github.io/PythonCertDevel/modules/Iteration.html#else
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/UWPCE-PythonCert-ClassRepos/Fall2018-PY210A/pull/58#pullrequestreview-169039822>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AqE_F7x6mteTTNxSkN3qSMMOJHCQTt7Mks5uo8G_gaJpZM4Xxa4f>
> .
>
